Cyclic OneToMany reference fails code generation

##### **Overview of the issue**

I want to generate code where a Person knows 0 to many other Persons. Without this relationship, code generates fine. As soon as I add this relationship, I get an error during Typescript code generation.

##### **Motivation for or Use Case**

I think it is legitimate to have a OneToMany relationship to the same entity. I considered making it a ManyToMany, but I think it will only complicate matters.


##### **Reproduce the error**

Import JDL into project containing this:

entity Person {
    description String,
    mboxSha1 String
}

Code above generates fine. Now, add:

relationship OneToMany {
	Person{knows} to Person
}

Code now gives errors:

 ERROR  Failed to compile with 1 errors                                                                                                                          15:50:34

 error  in /Users/jeroen/git/viervier/src/main/webapp/app/shared/model/person.model.ts

ERROR in /Users/jeroen/git/viervier/src/main/webapp/app/shared/model/person.model.ts(1,10):
TS2440: Import declaration conflicts with local declaration of 'IPerson'.

   1070 modules

ERROR in /Users/jeroen/git/viervier/src/main/webapp/app/shared/model/person.model.ts
ERROR in /Users/jeroen/git/viervier/src/main/webapp/app/shared/model/person.model.ts(1,10):
TS2440: Import declaration conflicts with local declaration of 'IPerson'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! viervier@0.0.1-SNAPSHOT webpack: `node --max_old_space_size=4096 node_modules/webpack/bin/webpack.js "--config" "webpack/webpack.dev.js" "--env.stats=minimal"`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the viervier@0.0.1-SNAPSHOT webpack script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jeroen/.npm/_logs/2020-04-05T13_50_34_367Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! viervier@0.0.1-SNAPSHOT webpack:build:main: `npm run webpack -- --config webpack/webpack.dev.js --env.stats=minimal`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the viervier@0.0.1-SNAPSHOT webpack:build:main script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jeroen/.npm/_logs/2020-04-05T13_50_34_606Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! viervier@0.0.1-SNAPSHOT webpack:build: `npm run cleanup && npm run webpack:build:main`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the viervier@0.0.1-SNAPSHOT webpack:build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jeroen/.npm/_logs/2020-04-05T13_50_34_642Z-debug.log





##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

6.8.0

##### **JHipster configuration**

INFO! Using JHipster version installed locally in current project's node_modules
INFO! Executing jhipster:info
INFO! Options: from-cli: true
##### **JHipster Version(s)**

```
viervier@0.0.1-SNAPSHOT /Users/jeroen/git/viervier
└── generator-jhipster@6.8.0 

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "nl.kransen.viervier",
      "nativeLanguage": "nl"
    },
    "jhipsterVersion": "6.8.0",
    "applicationType": "monolith",
    "baseName": "viervier",
    "packageName": "nl.kransen.viervier",
    "packageFolder": "nl/kransen/viervier",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "bXktc2VjcmV0LXRva2VuLXRvLWNoYW5nZS1pbi1wcm9kdWN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl",
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "react",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1586072794806,
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "nl",
    "languages": ["nl"],
    "blueprints": []
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Person {
  description String,
  mboxSha1 String
}
entity GeneratedResume {
  created LocalDate,
  resume AnyBlob
}
entity Skill {
  description String
}
entity Education {
  description String,
  type EducationType
}
entity Project {
  name String,
  company String,
  startDate LocalDate,
  endDate LocalDate
}
enum EducationType {
  UNIVERSITY,
  COURSE
}

relationship OneToMany {
  Person{resume} to GeneratedResume{person},
  Person{hasSkill} to Skill{person},
  Person{studied} to Education{person},
  Person{worked} to Project{person}
}

paginate Person, Project with pagination
paginate Skill with infinite-scroll
service Person, GeneratedResume, Skill, Education, Project with serviceImpl

</pre>
</details>


##### **Environment and Tools**

java version "11.0.6" 2020-01-14 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.6+8-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.6+8-LTS, mixed mode)

git version 2.14.3 (Apple Git-98)

node: v12.16.1

npm: 6.13.4

Docker version 19.03.5, build 633a0ea

docker-compose version 1.24.1, build 4667896b

INFO! Congratulations, JHipster execution is complete!


entity Person {
    description String,
    mboxSha1 String
}

entity GeneratedResume {
	created LocalDate,
    resume Blob
}

entity Skill {
    description String
}

entity Education {
    description String,
    type EducationType
}

enum EducationType {
    UNIVERSITY, COURSE
}

entity Project {
	name String,
    company String,
    startDate LocalDate,
    endDate LocalDate
}

relationship OneToMany {
	Person{knows} to Person
}

relationship OneToMany {
	Person{hasSkill} to Skill
}

relationship OneToMany {
	Person{studied} to Education
}

relationship OneToMany {
	Person{worked} to Project
}

relationship OneToMany {
	Person{resume} to GeneratedResume
}


// Set pagination options
paginate Person, Project with pagination
paginate Skill with infinite-scroll

// dto * with mapstruct

// Set service options to all except few
service all with serviceImpl
// Set an angular suffix
//angularSuffix * with mySuffix


##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

entity Person {
    description String,
    mboxSha1 String
}

relationship OneToMany {
        Person{knows} to Person
}


##### **Browsers and Operating System**

Not on browser, fails in terminal

-   [X] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/entity-client/templates/react/src/main/webapp/app/entities/entity.model.ts.ejs
+++ b/generators/entity-client/templates/react/src/main/webapp/app/entities/entity.model.ts.ejs
@@ -26,7 +26,9 @@ const enumImports = generateEntityClientEnumImports(fields);
 import { Moment } from 'moment';
 <%_ } _%>
 <%_ typeImports.forEach( (importedPath, importedType) => { _%>
+    <%_ if (importedType !== `I${entityReactName}`) { _%>
 import { <%- importedType %> } from '<%- importedPath %>';
+    <%_ } _%>
 <%_ }); _%>
 <%_ enumImports.forEach( (importedPath, importedType) => { _%>
 import { <%- importedType %> } from '<%- importedPath %>';

